feat: Editorial review: Long animation frames API docs (#33039)
mdn/content: files/en-us/web/api/performancelonganimationframetiming/blockingduration/index.md

- `duration` is a measure of the LoAF's total response time, which is useful for understanding whether the frame's layout, painting, etc. took a long time.
- `blockingDuration` is a measure of the total time the LoAF blocked the main thread from responding to high priority tasks, such as user interactions, which can cause the UI to feel [janky](/en-US/docs/Glossary/Jank). To put it another way, it is a measure of the effect the LoAF will have on responsiveness.

The reason why the `blockingDuration` of each task is calculated as `duration - 50ms` is that response delays of above 50ms start to be perceptible by users. This threshold is when users start noticing sluggishness; therefore, the time above the 50ms mark is important to measure for determining the severity of jank.

> **Note:** See [Total Blocking Time (TBT)](https://web.dev/articles/tbt) for more details.

## Value

A {{domxref("DOMHighResTimeStamp")}}.

## Examples

See [Long animation frame timing](/en-US/docs/Web/API/Performance_API/Long_animation_frame_timing#examples) for examples related to the Long Animation Frames API.

## Specifications

{{Specifications}}

## Browser compatibility

{{Compat}}

## See also

> - [Long animation frame timing](/en-US/docs/Web/API/Performance_API/Long_animation_frame_timing)

Author: I'm not sure I see the point of this; we don't add "guide" or "reference" to all the See also links habitually, do we? I've never done this before.

> - [Long animation frame timing](/en-US/docs/Web/API/Performance_API/Long_animation_frame_timing)

Reviewer: Correct, we don't add "reference", "property", or "data type" because the names already appear in code styling and we assume that to be indicative enough. Over in some CSS pages, we've wanted to, for example, distinguish between module pages and guide pages, that's where we've added the adjective. Honestly, it is a bit all over the place at the moment so we can omit "guide" here if that keeps the API pages similar.

- {{domxref("PerformanceScriptTiming")}}

> - [Optimize long tasks](https://web.dev/articles/optimize-long-tasks), web.dev (2024): contains more information on long tasks and yielding.

Author: ok, updated

> - [Optimize long tasks](https://web.dev/articles/optimize-long-tasks), web.dev (2024): contains more information on long tasks and yielding.

Reviewer: Thank you!

---
title: "PerformanceLongAnimationFrameTiming: blockingDuration property"
short-title: blockingDuration
slug: Web/API/PerformanceLongAnimationFrameTiming/blockingDuration
page-type: web-api-instance-property
status:
  - experimental
browser-compat: api.PerformanceLongAnimationFrameTiming.blockingDuration
---

{{SeeCompatTable}}{{APIRef("Performance API")}}

> The **`blockingDuration`** readonly property of the {{domxref("PerformanceLongAnimationFrameTiming")}} interface returns a {{domxref("DOMHighResTimeStamp")}} indicating the total time the main thread was blocked from responding to high priority tasks, such as user input, in milliseconds.

Reviewer 2: ```suggestion
The **`blockingDuration`** read-only property of the {{domxref("PerformanceLongAnimationFrameTiming")}} interface returns a {{domxref("DOMHighResTimeStamp")}} indicating the total time the main thread was blocked from responding to high priority tasks, such as user input, in milliseconds.
```

> The **`blockingDuration`** read-only property of the {{domxref("PerformanceLongAnimationFrameTiming")}} interface returns a {{domxref("DOMHighResTimeStamp")}} indicating the total time the main thread was blocked from responding to high priority tasks, such as user input, in milliseconds.

Author: updated same as in the interface index page

## Description

`blockingDuration` is calculated by taking all the [long tasks](/en-US/docs/Web/API/PerformanceLongTaskTiming#description) within the LoAF that have a `duration` of more than `50ms`, subtracting `50ms` from each, adding the rendering time to the longest task time, and summing the results. Let's look at an example to clarify what this means.

> Take the case of a JavaScript file that takes a total of 145ms to process. After the first major chunk of script, which is processed after 65ms, we could consider breaking the execution of the script into two tasks, the second one taking 80ms to execute. This is preferable to having it all executed in one task, as it gives the browser a chance to handle user interactions in between. This is also known as **yielding**, and could be done for example by inserting a {{domxref("setTimeout()")}} after the first major chunk of script.

Author: I liked this rewrite, so I pretty much used all of it verbatim. I updated "you can set up yielding" to "you can yield".

> > **Note:** See [Optimize long tasks](https://web.dev/articles/optimize-long-tasks) for more information on long tasks and yielding.

Author: OK, I've moved it to "See also" and added the latest update year in parens

> There are three options to consider here in how the script might end up being processed.

Author: Updated

1. If we yield after the first 65ms, the browser can decide to render a frame, then run the rest of the script
2. Alternatively, the browser could run the rest of the script first, then render the frame.

> 3. We could also decide **not** to yield and let the browser deal with the entire script in one task.

Author: These suggestions are nice — all used verbatim

> > **Note:** The browser will generally try to prioritize important tasks (such as user interactions) and rendering new frames over less important tasks it might have queued. The browser will _try_ to render a new frame every 16ms.

Author: like it; updated

> Assuming the rendering time is 10ms, timings for the LoAFs in each of the three options are as follows:

Author: I ended up with:

> We mentioned earlier that the total processing time for the script is 145ms. Assuming the time for rendering the UI update is 10ms, the timings for the LoAFs in each of the three options are as follows:

I wanted to be specific that we are talking about processing the script, and I also put "rendering the UI update", as technically in the first version we have two frames, plus I'm not sure if there is other things implied in "the rendering"

> Assuming the rendering time is 10ms, timings for the LoAFs in each of the three options are as follows:

Reviewer: Looks good

| Option | `duration` (LoAF 1) | `blockingDuration` (LoAF1)        | `duration` (LoAF2) | `blockingDuration` (LoAF2) |
| ------ | ------------------- | --------------------------------- | ------------------ | -------------------------- |
| 1.     | 65ms                | 15ms (65 - 50)                    | 80ms               | 40ms (80 + 10 - 50)        |
| 2.     | 145ms (65 + 80)     | 55ms ((65 - 50) + (80 + 10 - 50)) | n/a\*              | n/a\*                      |
| 3.     | 145ms (65 + 80)     | 105ms ((65 + 80) + 10 - 50)       | n/a\*              | n/a\*                      |

> `*` In options 2. and 3., there is only a single LoAF.

Reviewer: ```suggestion
`*` In options 2 and 3, there is only a single LoAF.
```

Author: Updated. I also decided to remove the periods from the table's leftmost column, as it looks neater that way.

Reviewer: Agree

> Note that the total `duration` in all three instances is 145ms. The total `blockingDuration` in the first two is the same (55ms) — the browser just decides to split the work in different ways in each.

Reviewer: Oh, I've repeated the `145ms` info before the table - I found it useful as a reminder because it was last mentioned at the beginning of the example

Author: That's not a big problem; I've updated this sentence to:

> Note that the total `blockingDuration` in the first two is the same (55ms) — in each case the browser has decided to split the work in different ways.

> The third option however has a much longer `blockingDuration` because the browser is completely blocked and unable to interrupt the long task at all. This highlights the importance of optimizing long tasks by yielding — regardless of how the browser decides to handle the tasks, the blocking duration will still be less than if you don't yield at all.

Reviewer: ```suggestion
Option 3, however, has a much longer `blockingDuration` because the browser is completely blocked and unable to interrupt the long task at all. This highlights the importance of optimizing long tasks by yielding — regardless of how the browser decides to handle the tasks, the blocking duration will still be less than if you don't yield at all.
```

Author: updated

The difference between `duration` and `blockingDuration` of LoAFs can be summarised as follows:

- `duration` is a measure of the LoAF's total response time, which is useful for understanding whether the frame's layout, painting, etc. took a long time.
- `blockingDuration` is a measure of the total time the LoAF blocked the main thread from responding to high priority tasks, such as user interactions, which can cause the UI to feel [janky](/en-US/docs/Glossary/Jank). To put it another way, it is a measure of the effect the LoAF will have on responsiveness.

> The reason why the `blockingDuration` of each task is calculated as `duration - 50ms` is that response delays of above 50ms start to be perceptible by users. This threshold is when users start noticing sluggishness; therefore, it is the time above the 50ms that is important to measure for determining the severity of jank.

Reviewer: ```suggestion
The reason why the `blockingDuration` of each task is calculated as `duration - 50ms` is that response delays of above 50ms start to be perceptible by users. This threshold is when users start noticing sluggishness; therefore, it is the time above the 50ms mark that is important to measure for determining the severity of jank.
```

Author: Added; I also decided to prune the sentence a bit more to make it a bit less wordy:

> ...therefore, the time above the 50ms mark is important to measure for determining the severity of jank.

Reviewer: Yeah, this is better

> > **Note:** See [Total Blocking Time (TBT)](https://web.dev/articles/tbt) for more details.

Reviewer: Can be moved to "See also"?

Author: I'd rather leave this one here. I feel like it is very context-dependent — putting it down in "See also" will then require a whole bunch of explaining to clarify what it is referring to.

Reviewer: Got it. Just to clarify, we don't need to provide any description of the content on the target page in 'See also'.
If you keep this external link here, maybe it does not need to be a "Note"?

Reviewer: Yeah

## Value

A {{domxref("DOMHighResTimeStamp")}}.

## Examples

> See [Long animation frame timing](/en-US/docs/Web/API/Performance_API/Long_animation_frame_timing) for Long Animation Frames API examples.

Reviewer: same comment as https://github.com/mdn/content/pull/33039#discussion_r1567717604

Author: I've gone through and updated all instances of this sentence in the PR to be the same as in the interface landing page.

Reviewer: thank you!
